### PR TITLE
PIMS-977 Adding /api/admin/users endpoint.

### PR DESCRIPTION
**Pull request by Fosol**
Also updated the API authentication/authorization error handling to be consistent with all our other error handling.
Created semi-generic implementation for submitting filtered queries from an endpoint to the DAL.

### CODE REVIEW

File: backend/api/Startup.cs [hunk 1]
```diff
@@ -98,13 +98,8 @@ public void ConfigureServices(IServiceCollection services)
...
                         {
                             context.NoResult();
-                            context.Response.StatusCode = 500;
-                            context.Response.ContentType = "text/plain";
-                            if (Environment.IsDevelopment())
-                            {
-                                return context.Response.WriteAsync(context.Exception.ToString());
-                            }
-                            return context.Response.WriteAsync("An error occurred processing your authentication.");
+                            context.Response.StatusCode = 401;
```
Comment (Fosol) [inline backend/api/Startup.cs]: Authentication errors will now use the error handling middleware and return 401.

File: backend/api/Startup.cs [hunk 2]
```diff
@@ -163,14 +159,8 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
             if (!env.IsProduction())
             {
                 app.UseDatabaseErrorPage();
-                //app.UseDeveloperExceptionPage();
-
                 UpdateDatabase(app);
             }
-            else
```
Comment (Fosol) [inline backend/api/Startup.cs]: Clean up

File: backend/dal/Helpers/Extensions/QueryableExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+        /// <summary>
+        /// Order the query results by the specified property names.
+        /// Each property can also specify the direction of the sort (i.e. "Name asc", "Name ascending", "Name desc", "Name descending").
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="source"></param>
+        /// <param name="propertyName"></param>
+        /// <returns></returns>
+        public static IQueryable<T> OrderByProperty<T>(this IQueryable<T> source, params string[] propertyName)
```
Comment (Fosol) [inline backend/dal/Helpers/Extensions/QueryableExtensions.cs]: Provides a way to dynamically handle sorting by an array of property path values.

File: backend/dal/Helpers/Extensions/QueryableExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+        /// <summary>
+        /// Order the query results by the specified property names.
+        /// Any property name that doesn't exist will be ignored.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="source"></param>
+        /// <param name="propertyName"></param>
+        /// <returns></returns>
+        public static IQueryable<T> OrderByPropertyDescending<T>(this IQueryable<T> source, params string[] propertyName)
```
Comment (Fosol) [inline backend/dal/Helpers/Extensions/QueryableExtensions.cs]: Provides a way to dynamically handle sorting by an array of property path values.

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 5]
```diff
@@ -31,18 +31,48 @@ public class UserService : BaseService<User>, IUserService
...
         /// <summary>
         /// Get a page of users from the datasource.
+        /// The filter will allow queries to search for anything that starts with the following properties; DisplayName, FirstName, LastName, Email, Agencies.
         /// </summary>
         /// <param name="page"></param>
         /// <param name="quantity"></param>
-        /// <param name="sort"></param>
         /// <returns></returns>
-        public Paged<User> GetNoTracking(int page = 1, int quantity = 10, string sort = null)
+        public Paged<User> GetNoTracking(int page = 1, int quantity = 10)
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: Basic method to page through all users.

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 6]
```diff
@@ -31,18 +31,48 @@ public class UserService : BaseService<User>, IUserService
...
+            return GetNoTracking(new UserFilter(page, quantity));
+        }
+
+        /// <summary>
+        /// Get a page of users from the datasource.
+        /// The filter will allow queries to search for the following property values; DisplayName, FirstName, LastName, Email, Agencies.
+        /// </summary>
+        /// <param name="filter"></param>
+        /// <returns></returns>
+        public Paged<User> GetNoTracking(UserFilter filter = null)
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: Advanced method to page through filtered users.

File: backend/entities/Models/UserFilter.cs [hunk 7]
```diff
@@ -0,0 +1,129 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Pims.Dal.Entities.Models
+{
+    /// <summary>
+    /// UserFilter class, provides a model for filtering user queries.
+    /// </summary>
+    public class UserFilter
```
Comment (Fosol) [inline backend/entities/Models/UserFilter.cs]: Provides a strongly typed list of what we want to allow users to filter/search/sort on.

File: backend/api/Areas/Admin/Controllers/UserController.cs [hunk 8]
```diff
@@ -53,18 +55,27 @@ public UserController(ILogger<UserController> logger, IConfiguration configurati
...
-            if (quantity > 50) quantity = 50;
+            var uri = new Uri(this.Request.GetDisplayUrl());
+            var query = Microsoft.AspNetCore.WebUtilities.QueryHelpers.ParseQuery(uri.Query);
+            _logger.LogDebug(query.ToString());
+            return GetUsers(new EModel.UserFilter(query));
+        }
 
-            var results = _pimsAdminService.User.GetNoTracking(page, quantity, sort);
+        /// <summary>
+        /// GET - Returns a paged array of users from the datasource.
```
Comment (devinleighsmith) [inline backend/api/Areas/Admin/Controllers/UserController.cs]: POST?

Comment (Fosol) [reply backend/api/Areas/Admin/Controllers/UserController.cs]: Gives us two ways to make a filtered request for users.  We can pass them as query string parameters in the GET or in a JSON body in a POST.

File: backend/dal/Helpers/Extensions/QueryableExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+                    var orderExpression = convertMethod.Invoke(null, new[] { parts[0] });
+                    if (orderExpression != null)
+                    {
+                        var genericMethod = OrderByMethod.MakeGenericMethod(typeof(T), rType);
+                        query = (IQueryable<T>)genericMethod.Invoke(null, new[] { query, orderExpression });
+                    }
+                    else
+                    {
+                        if (!PropertyExists<T>(parts[0])) continue;
+                        ParameterExpression paramterExpression = Expression.Parameter(typeof(T));
```
Comment (devinleighsmith) [inline backend/dal/Helpers/Extensions/QueryableExtensions.cs]: /nit paramter

Comment (Fosol) [reply backend/dal/Helpers/Extensions/QueryableExtensions.cs]: fixed

File: backend/api/Areas/Admin/Controllers/UserController.cs (hunk 8 above)
Comment (devinleighsmith) [reply backend/api/Areas/Admin/Controllers/UserController.cs]: I just meant the comment seemed wrong.

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 10]
```diff
@@ -31,18 +31,48 @@ public class UserService : BaseService<User>, IUserService
...
+            if (filter.Page < 1) filter.Page = 1;
+            if (filter.Quantity < 1) filter.Quantity = 1;
+            if (filter.Quantity > 50) filter.Quantity = 50;
+
             var query = this.Context.Users.AsNoTracking();
-            var users = query.Skip((page - 1) * quantity).Take(quantity);
-            return new Paged<User>(users.ToArray(), page, quantity, query.Count());
+            if (filter != null)
+            {
+                if (!string.IsNullOrWhiteSpace(filter.DisplayName))
```
Comment (devinleighsmith) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: This approach looks good but doesn't seem to scale well with large numbers of fields. Thoughts?

Comment (Fosol) [reply backend/dal/Services/Admin/Concrete/UserService.cs]: Definitely it's primary flaw.  While it allows us control over what is queryable, it makes us have to work harder for complex options.  I can rewrite using reflection at some point, but for now it doesn't make sense to do this.

File: backend/api/Areas/Admin/Controllers/UserController.cs [hunk 11]
```diff
@@ -53,18 +55,27 @@ public UserController(ILogger<UserController> logger, IConfiguration configurati
...
+        /// </summary>
+        /// <param name="filter"></param>
+        /// <returns>Paged object with an array of users.</returns>
+        [HttpPost("/api/admin/users")]
+        public IActionResult GetUsers(EModel.UserFilter filter)
```
[the remark this answers is not in the document]
Comment (asanchezr) [reply backend/api/Areas/Admin/Controllers/UserController.cs]: If we use the POST here for a "get/read" operation. Then what's the URL for adding a user?

[the remark this answers is not in the document]
Comment (Fosol) [reply backend/api/Areas/Admin/Controllers/UserController.cs]: This leads us back to the conversation you started earlier about standardizing our routes.  I think it's worthwhile to push this to the dev channel and get a consensus.  I'm fine with going either way, as there are pros and cons.

File: backend/dal/Services/Admin/Concrete/UserService.cs (hunk 5 above)
Comment (asanchezr) [reply backend/dal/Services/Admin/Concrete/UserService.cs]: Could you please add a comment why we are no longer going to support sorting here? thanks

Comment (Fosol) [reply backend/dal/Services/Admin/Concrete/UserService.cs]: The more advanced method is a few lines down.  All this method provides is a simple interface for requesting a page of all users.  The method probably should just be removed, but I don't think it's critical atm.

**Pull request merged**